File: python/atelier_bench/services/part_usage_service.py
"""part-usage service."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..support.abstract_service import AbstractService

if TYPE_CHECKING:
    from ..dataset import Dataset


class PartUsageService(AbstractService):
    NAME = "part-usage"

    def __init__(self) -> None:
        super().__init__(self.NAME)

    def consumed_units(self, data: "Dataset") -> int:
        return sum(part.consumed_quantity() for part in data.parts)

File: go/atelier/services/part_usage_service.go
// Package services carries the 12 services. Note the name: core/services is a
// DIFFERENT package with the same name, and a file needing both aliases one.
package services

import (
	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/dataset"
	"atelier.example/lane/atelier/support"
)

// Compile-time proof of satisfaction.
var _ contracts.ServiceContract = (*PartUsageService)(nil)

// PartUsageService answers the part-usage questions.
type PartUsageService struct {
	support.BaseService
}

// NewPartUsageService builds the service.
func NewPartUsageService() *PartUsageService {
	return &PartUsageService{BaseService: support.NewBaseService("part-usage")}
}

// ConsumedUnits is how many part units left the shelf.
func (s *PartUsageService) ConsumedUnits(data *dataset.Dataset) int {
	units := 0
	for _, part := range data.Parts {
		units += part.ConsumedQuantity()
	}
	return units
}

// Run renders the service result as report rows.
func (s *PartUsageService) Run(data *dataset.Dataset) []contracts.ReportRow {
	return []contracts.ReportRow{contracts.Row("units", float64(s.ConsumedUnits(data)))}
}
